aten/src/ATen/Dispatch.h

```diff
@@ -284,6 +310,25 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                                                          \
   }()
 
+// This macro should be used to dispatch bfloat16 on ROCm.
+// On non-ROCm platforms, it ignores bfloat16 data type.
+#define AT_DISPATCH_ALL_TYPES_AND_BFLOAT16_AND(SCALARTYPE, TYPE, NAME, ...)                             \
```

Comment:
and this one???

Cc'ing @gchanan to take another look at these macros, I can never remember the rules for when we're allowed to add more.

Reply:
and this one to `AT_DISPATCH_ALL_TYPES_AND_ROCM_BFLOAT16_AND`?

Reply:
I guess my first question is why we need this?  Do the kernels not handle bfloat16?  I would have expected this sort of thing for say, cudNN, where the version of the library might not support it, but for generic ops like below this is surprising to me.

Reply:
As for the dispatch macro itself, the restriction @ezyang is probably thinking of is adding more types to an existing macro, because this would impact downstream users.

Assuming we need to conditionally compile things (see above), I don't think this is the best way to do it because it complicates the dispatch macro story (now people are going to wonder why we don't have more AND_BFLOAT16 macros instead of just using one of the AND( macros.  Also, this mismatch between RoCM and CUDA should go away, so there's no need long term for this.

How about this instead of:
```
AT_DISPATCH_FLOATING_TYPES_AND_BFLOAT16_AND(at::ScalarType::Half, iter.dtype(), "op_name", [&]() {
          ...
})
```

we do something like (you'll have to clean this up):
```
AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::BFloat16, at::ScalarType::Half, iter.dtype(), "op_name", [&]() {
#ifndef __HIP_PLATFORM_HCC && std::is_same<scalar_t, bfloat16>
    // See Note: [BFloat16 ROCm/CUDA]
    AT_ERROR(#NAME, " not implemented for '", toString(TYPE), "'");
#else
   ...
#endif
})
````

you can put that structure in a macro (AT_SKIP_BFLOAT16_IF_NOT_ROCM) if you want, just have it live in a ROCm or BFloat16 specific file instead of a dispatch-wide file.

Reply:
@gchanan The reason I added this is to compile BFloat16 only on ROCm. I initially used `AT_DISPATCH_FLOATING_TYPES_AND2` macro to enable on both CUDA and ROCm, but that led to failures on CUDA. Hence, I refactored the PR using this new macro to compile and test only ROCm. The idea was that, this new macro is a short term solution and should be unified with 'AT_DISPATCH_FLOATING_TYPES_AND2` when CUDA also works correctly.

Concerning using `#ifdef` within the existing `AT_DISPATCH_FLOATING_TYPES_AND2` macro, I can change the PR to use this approach but IMHO this may lead to relatively more code changes. Also, is there any suggestion on where to define `AT_SKIP_BFLOAT16_IF_NOT_ROCM` type of macro? I don't think defining it in a ROCm specific file would work since it would be needed on CUDA as well.